### PR TITLE
feat: add close functionalitiy for dialog and sheet components

### LINKED ISSUE #88: Feature: Close Functionality for Dialog Component
Right now there's no such wrapper code available in the docs for close functionality in Dialog component. I have tried to implement the same and it's working all fine.

```tsx
const DialogClose = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Close>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Close>
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Close
    ref={ref}
    {...props}
  />
))
DialogClose.displayName = DialogPrimitive.Title.displayName
```

This can be used like the snippet shown below
```tsx
<Dialog>
  <DialogTrigger>Open</DialogTrigger>
  <DialogContent>
    <DialogHeader>
      <DialogTitle>Are you sure absolutely sure?</DialogTitle>
      <DialogDescription>
        This action cannot be undone. This will permanently delete your account
        and remove your data from our servers.
      </DialogDescription>
    </DialogHeader>
    <DialogFooter>
      {/* this is how it can be used */}
      <DialogClose />
    </DialogFooter>
  </DialogContent>
</Dialog>
```

Hope this helps! Happy to have a chat around this issue/feature-request.

### PR DESCRIPTION
**Pull request by keon**
resolves #88

### CODE REVIEW

File: templates/next-template/components/ui/dialog.tsx [hunk 1]
```diff
@@ -119,6 +119,17 @@ const DialogDescription = React.forwardRef<
 ))
 DialogDescription.displayName = DialogPrimitive.Description.displayName
 
+const DialogClose = React.forwardRef<
```
Comment (its-monotype) [inline templates/next-template/components/ui/dialog.tsx:122]: Why not just const DialogClose = DialogPrimitive.Close

**Comment by shadcn** [comment]
We already have this. You can see it at https://ui.shadcn.com/docs/primitives/dialog

zzaOuOqf

However, this has been reported by a few other people. So I wonder why it's not showing up. I'll take a look at it.

File: templates/next-template/components/ui/dialog.tsx (hunk 1 above)
Comment (keon) [reply templates/next-template/components/ui/dialog.tsx]: simplified it!

**Comment by keon** [comment]
> We already have this. You can see it at https://ui.shadcn.com/docs/primitives/dialog
> 
> zzaOuOqf
> 
> However, this has been reported by a few other people. So I wonder why it's not showing up. I'll take a look at it.

@shadcn Yes the x button on the top right works - and it is not the issue here.
DialogClose is still needed to create any close button in the content.

**Comment by NikAtNight** [comment]
Slight issue outside of this.

If you pass a prop to open on `<Dialog open={open}>

we also need something like onClose then.

The x at the top stops working and clicking outside the dialog also stops working.

But with the addition of DialogClose this would solve my problem.

**Comment by Mohsinsiddi** [comment]
> Slight issue on the topic of Closing the Dialog.
> 
> If you pass the prop open on `<Dialog open={open}>`
> 
> we also need something like onClose then.
> 
> The x at the top stops working and clicking outside the dialog also stops working.
> 
> But with the addition of DialogClose this would solve my problem if the submit button were to auto trigger a close.

I hope vercel team will provide a update on this son

**Comment by eliuddyn** [comment]
> But with the addition of DialogClose this would solve my problem if the submit button were to auto trigger a close.

According to [Radix-UI](https://www.radix-ui.com/docs/primitives/components/dialog#close-after-asynchronous-form-submission):

`import React from 'react';
import * as Dialog from '@radix-ui/react-dialog';

const wait = () => new Promise((resolve) => setTimeout(resolve, 1000));

export default () => {
  const [open, setOpen] = React.useState(false);

  return (
    <Dialog.Root open={open} onOpenChange={setOpen}>
      <Dialog.Trigger>Open</Dialog.Trigger>
      <Dialog.Portal>
        <Dialog.Overlay />
        <Dialog.Content>
          <form
            onSubmit={(event) => {
              wait().then(() => setOpen(false));
              event.preventDefault();
            }}
          >
            {/** some inputs */}
            <button type="submit">Submit</button>
          </form>
        </Dialog.Content>
      </Dialog.Portal>
    </Dialog.Root>
  );
};`

For the Sheet component should be:

`<Sheet open={open} onOpenChange={setOpen}>
            ...
</Sheet>`

**Comment by faqi** [comment]
> > But with the addition of DialogClose this would solve my problem if the submit button were to auto trigger a close.
> 
> According to [Radix-UI](https://www.radix-ui.com/docs/primitives/components/dialog#close-after-asynchronous-form-submission):
> 
> `import React from 'react'; import * as Dialog from '@radix-ui/react-dialog';
> 
> const wait = () => new Promise((resolve) => setTimeout(resolve, 1000));
> 
> export default () => { const [open, setOpen] = React.useState(false);
> 
> return ( <Dialog.Root open={open} onOpenChange={setOpen}> <Dialog.Trigger>Open</Dialog.Trigger> <Dialog.Portal> <Dialog.Overlay /> <Dialog.Content> <form onSubmit={(event) => { wait().then(() => setOpen(false)); event.preventDefault(); }} > {/** some inputs */} Submit </Dialog.Content> </Dialog.Portal> </Dialog.Root> ); };`
> 
> For the Sheet component should be:
> 
> `<Sheet open={open} onOpenChange={setOpen}> ... </Sheet>`

...................
Thanks. it worked for me

**Comment by NikAtNight** [comment]
I've recently started to work on some of my old projects and I've just done this to my `DialogContent` component in the auto generated `dialog.tsx` file.

This will allow the popup to close now, but does not solve closing the dialog when clicking outside of the popup.
```
// Update the forwardRef and added a handleClose
const DialogContent = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Content> & {
    handleClose?: () => void;
  }
>(({ className, children, handleClose, ...props }, ref) => ( //added here as a prop
  <DialogPortal>
    <DialogOverlay />
    <DialogPrimitive.Content
      ref={ref}
      className={cn(
        "fixed left-[50%] top-[50%] z-50 grid w-full max-w-lg translate-x-[-50%] translate-y-[-50%] gap-4 border bg-background p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg md:w-full",
        className
      )}
      {...props}
    >
      {children}
      <DialogPrimitive.Close
        className="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground"
        onClick={handleClose} // added onClick handler
      >
        <Cross2Icon className="h-4 w-4" />
        <span className="sr-only">Close</span>
      </DialogPrimitive.Close>
    </DialogPrimitive.Content>
  </DialogPortal>
));
DialogContent.displayName = DialogPrimitive.Content.displayName;
```

**Comment by chukwudinweze** [comment]
> > Slight issue on the topic of Closing the Dialog.
> > If you pass the prop open on `<Dialog open={open}>`
> > we also need something like onClose then.
> > The x at the top stops working and clicking outside the dialog also stops working.
> > But with the addition of DialogClose this would solve my problem if the submit button were to auto trigger a close.
> 
> I hope vercel team will provide a update on this son

Many thanks. This worked for me

**Comment by joaom00** [comment]
> This will allow the popup to close now, but does not solve closing the dialog when clicking outside of the popup.
> 
> I think these changes should be added to the PR both for Dialog & Sheet

Not sure I understand sorry. but if you use like:
```tsx
const [open, setOpen] = React.useState(false)

<Dialog open={open} onOpenChange={setOpen}>
  ...
</Dialog>
```
the `X` at the top and clicking outside still doesn't work?

**Comment by NikAtNight** [comment]
> > This will allow the popup to close now, but does not solve closing the dialog when clicking outside of the popup.
> > I think these changes should be added to the PR both for Dialog & Sheet
> 
> Not sure I understand sorry. but if you use like:
> 
> ```tsx
> const [open, setOpen] = React.useState(false)
> 
> <Dialog open={open} onOpenChange={setOpen}>
>   ...
> </Dialog>
> ```
> 
> the `X` at the top and clicking outside still doesn't work?

Here is what you can do. Since the onClose is a part of the `DialogContent` you need to pass it here
```
<Dialog open={visible}>
  <DialogTrigger asChild></DialogTrigger>
  <DialogContent onClose={() => setVisible(false)}>
   ...
  </DialogContent>
</Dialog>
```

**Comment by joaom00** [comment]
> > > This will allow the popup to close now, but does not solve closing the dialog when clicking outside of the popup.
> > > I think these changes should be added to the PR both for Dialog & Sheet
> > 
> > 
> > Not sure I understand sorry. but if you use like:
> > ```tsx
> > const [open, setOpen] = React.useState(false)
> > 
> > <Dialog open={open} onOpenChange={setOpen}>
> >   ...
> > </Dialog>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > the `X` at the top and clicking outside still doesn't work?
> 
> If you update `dialog.tsx` as I did, here is what you can do. Since the onClose is a part of the `DialogContent` you need to pass it here
> 
> ```
> <Dialog open={open}>
>   <DialogTrigger asChild>
>   ...
>   </DialogTrigger>
>   <DialogContent onClose={() => setOpen(false)}>
>    ...
>   </DialogContent>
> </Dialog>
> ```
> 
> Or if you followed the other example refer to that code.

but if you pass `onOpenChange={setOpen}` to `<Dialog />` the `X` and clicking outside will work.

**Comment by NikAtNight** [comment]
@joaom00 Thanks for that! I misunderstood the first set of code then.

**Comment by shadcn** [comment]
I fixed it and added examples on how to use it do the docs. See https://github.com/shadcn-ui/ui/pull/1753 

Thanks everyone.

**Pull request closed without merging**